docs/CONFIG.md
# Chorus Configuration

The chorus binary requires one command line parameter which specifies the config file path.

The config file must be in TOML format. See the [TOML documentation](https://github.com/toml-lang/toml)

## Configuration Variables

### data_directory

This is the directory where chorus stores data.

Default is "/opt/chorus/var/chorus".

If deployed according to [docs/DEPLOYING.md](docs/DEPLOYING.md), is "/opt/chorus/var/chorus".

### ip_address

This is the IP address that chorus listens on. If deployed directly on the Internet, this should
be an Internet globally accessible IP address. If proxied or if testing locally, this can be
a localhost address.

Default is "127.0.0.1".

### port

This is the port that chorus listens on. If deployed directly on the Internet, this should
probably be 443 which is the expected default port for the "wss://" protocol.

Default is 443.

### hostname

This is the DNS hostname of your relay. This is used for verifying AUTH events, which specify
your relay host name.

Default is localhost

### chorus_is_behind_a_proxy

If chorus is behind a proxy like nginx, set this to true. In this case chorus will look for and
trust the `X-Real-Ip` HTTP request header to get the real IP of the client. This header MUST exist
or the connection will not be served.

Default is false.

### base_url

If chorus is behing a proxy, it can't compute it's Internet-visible URL. So set it here.
This is used currently by web management and blossom and maybe more in the future.

Default is not set.

### use_tls

If true, chorus will handle TLS, running over HTTPS.  If false, chorus run over HTTP.

If you are proxying via nginx, normally you will set this to false and allow nginx to handle
TLS.

Default is true

### certchain_pem_path

This is the path to your TLS certificate chain file.

If `use_tls` is false, this value is irrelevant.

Default is "/opt/chorus/etc/tls/fullchain.pem"

If deployed according to [docs/DEPLOYING.md](docs/DEPLOYING.md) using the direct method,
systemd service copies letsencrypt TLS certificates into this position on start.

### key_pem_path

This is the path to your TLS private key file.

If `use_tls` is false, this value is irrelevant.

Default is "/opt/chorus/etc/tls/privkey.pem"

If deployed according to [docs/DEPLOYING.md](docs/DEPLOYING.md) using the direct method,
systemd service copies letsencrypt TLS certificates into this position on start.

### name

This is an optional name for your relay, displayed in the NIP-11 response.

Default is "Chorus Default"

### description

This is an optional description for your relay, displayed in the NIP-11 response.

Default is "A default config of the Chorus relay".

### banner_url

This is an optional URL for an graphical banner representing your relay, displayed in the NIP-11 response.

### icon_url

This is an optional URL for an graphical icon representing your relay, displayed in the NIP-11 response.

### privacy_policy

This is an optional privacy policy as a blob of text (not a URL, not HTML).

### terms_of_service

This is an optional terms of service document as a blob of text (not a URL, not HTML).

### contact

This is an optional administrative contact for your relay, displayed in the NIP-11 response.

Default is None.

### contact_public_key_hex

This is an optional public key (hex format) for your relay's administrative contact, displayed in the NIP-11 response.

Deprecated "public_key_hex" also works.

Default is None.

### open_relay

If open_relay true, the relay behaves as an open public relay.

Default is false.

### admin_hex_keys

These are the public keys (hex format) of your relay's administrators. This does NOT automatically make them a relay user, but it will eventually allow them to add/remove users and moderators.

Default is `[]`

### verify_events

This is a boolean indicating whether or not chorus verifies incoming events.

This setting only skips verification of events that are submitted by AUTHed and authorized users. Chorus always verifies incoming AUTH events, and any event that is not submitted by an AUTHed and authorized relay user.

Default is true.

### allow_scraping

This is a boolean indicating whether or not scraping is allowed. Scraping is any filter where all of the following are true:

- `ids` is missing or empty
- `authors` is missing or empty
- There are no `#X` tag filters

Filter that fail to match these conditions will be rejected if `allow_scraping` is false.

If `allow_scraping` is true, be aware that filters that don't match any of these conditions (and are not bound by since/until/limit) have no indexes to speed up their query, so they scan through every single event on the relay.

See also `allow_scrape_if_limited_to` and # `allow_scrape_if_max_seconds`.

Default is false.

### allow_scrape_if_limited_to

This is a u32 count of events indicating a filter `limit` value under which a scrape is allowed, irrespective of the `allow_scraping` setting. Such scrapes are not expensive due to the limit. The limit must be specified in the filter; it does not do the scrape and then count to see if it was under the limit.

See `allow_scraping` to learn the definition of a scrape.

The default is 100.

### allow_scrape_if_max_seconds

This is a u64 number of seconds indicating a filter time range under which a scrape is allowed, irrespective of the `allow_scraping` setting. Such scrapes are rarely expensive due to the short time period.

See `allow_scraping` to learn the definition of a scrape.

The default is 7200.

### allow_scrape_if_negentropy

Within negentropy synchronization, allow scraping.

The default is true

### max_subscriptions

This is an integer indicating the maximum number of subscriptions a connection can have open at a given time.

If you set this too low, clients will be incentivised to resubmit updated subscriptions which will pull down the same events over again, instead of submitting a new subscription that only gets the additional events that the client wants. It may seem intuitive that setting this to a low value like 10 will decrease server load, but it will probably increase server load.

It is strongly recommended to not go below 16.

Default is 128.

### serve_ephemeral

Whether or not to accept and serve all ephemeral events to everybody.

Default is true.

### serve_relay_lists

Whether or not to accept and serve kind 10002 Relay List Metadata (NIP-65) events to everybody.

Default is true.

### server_log_level

How verbose to log issues with the main server code.

Possible values are: Trace, Debug, Info, Warn, Error

Default is Info

### library_log_level

How verbose to log library issues and other general issues

Possible values are: Trace, Debug, Info, Warn, Error

Default is Info

### client_log_level

How verbose to log issues with client requests

Possible values are: Trace, Debug, Info, Warn, Error

Default is Info

### enable_ip_blocking

Whether to block incoming connections based on recent prior behavior

Chorus normally blocks IP addresses for a short period preventing quick reconnections, and for a longer period if the previous connection ended in some error condition.

By setting this variable to false, it will allow all connections, incluing poorly behaving clients that reconnect over and over in a tight loop.

Default is true

### minimum_ban_seconds

Number of seconds to ban an IP address after disconnection.

Enforcing this minimum ban prevents clients from immediately reconnecting which can cause tight loops.

Only relevant if enable_ip_blocking is true.

Default is 1

### timeout_seconds

Number of seconds beyond which chorus times out a client that has no open subscriptions.

Default is 60

### max_connections_per_ip

Maximum number of websocket connections per IP address

Default is 5

### throttling_bytes_per_second

The maximum rate (excluding bursts) of data that will be transmitted over a websocket connection
(both directions, per connection). Beyond this rate (in a sustained way) the connection will be
closed.

Default is 1024576 bytes per second.

### throttling_burst

The allowable bursts of data beyond the normal rate.

We keep a count of how many bytes they are allowed, and that count starts at this number.
As bytes are consumed the count goes down, but we refund throttling_bytes_per_second every
second. If that bucket doesn't have enough, the burst won't be allowed and the connection
will be closed.

Default is 16777216 bytes.

### blossom_directory

Blossom server directory

Set to a filesystem directory where you want chorus to store files.

Blossom allows clients to upload files making them available for the public to download.
Our implementation makes all files publicly readable, but only chorus users can upload
or delete.  See https://github.com/hzrd149/blossom

Default is None

### enable_negentropy

Whether to allow negentropy syncing or not

If this is enabled, keep in mind that some negentropy syncs are scrapes and if you want to
allow those, you should also enable allow_scraping. But this will require scans of the entire
database since scrapes have no indexes.

Default is false


CHANGELOG.md
# v2.0.0

- IMPORTANT: You need to manually [Migrate](docs/MIGRATION.md) your users and moderators.
- Management commands have been added:  listadmins, listmoderators, grantmoderator, revokemoderator,
  listusers, grantuser, revokeuser, stats, listeventsneedingmoderation, clearevent, clearpubkey,
  listrole, grantrole, revokerole, fetchbannedevents
- fix: subscriptions will be CLOSED: auth-required if any matching event requires auth. Chorus
  will serve the redacted results first, but will not EOSE.
- fix: subscriptions will be CLOSED when completed, without EOSE, if the filter has any ids set.

# v1.7.2

- Support for NIP-62 (PR #1256) Right to Vanish
- Fix NIP-86 management request header determination
- Management adds: numconnections, uptime
- chorus_cmd: fetch_by_id

# v1.7.1

- Config setting allow_scrape_if_negentropy=true
- CLOSED no longer sent after CLOSE

# v1.7.0

- Negentropy support (you must set enable_negentropy=true)
- Blossom extensions are better

# v1.6.1

- base_url added (you should set this if you are behind a proxy like nginx)
- NIP-45 COUNT support
- NIP-45 PR #1561 HyperLogLog count support

# v1.6.0

- Support for running as a blossom server
- Websocket close frames provided on close
- Better pre-flight checks
- server-wide OPTIONS requests now handled

# v1.5.3

- Performance improvements (streaming body)
- Approval not required for kinds 0, 3, and reactions (as well as prior list)
- FIX honor enable_ip_blocking = false
- NEW `chorus_cmd` with `delete_by_id` and `delete_by_pubkey`
- optional `icon_url` for NIP-11

# version 1.5.2 (2024-07-31, d8689540)

- NEW CONFIG: `throttling_bytes_per_second` how many bytes are allowed per second
- NEW CONFIG: `throttling_burst` how many bytes can be accepted per connection in a burst
- Don't wait more than 5 seconds for connections to close when shutting down

# version 1.5.1 (2024-07-14, c9c71311)

- FIX: large non-utf8 messages were attempted to be logged causing a panic
- FIX: parse errors were not being punished so a nasty client could do a DoS of chorus with
  simple junk.

# version 1.5.0 (2024-07-13, 870e470d)

- BREAKING: If you run chorus behind a proxy like nginx, you MUST set the new `chorus_is_behind_a_proxy`
  config variable to true, and your proxy MUST set the `X-Real-IP` header.  If the header is missing,
  connections will not be served. If you fail to set the `chorus_is_behind_a_proxy` setting, the proxy
  IP address will be used directly, generally causing all connections to quickly become banned due to
  the bad behavior of just one client, or due to too many connections from a single IP.
- NEW CONFIG: `chorus_is_behind_a_proxy` (please set to true or false)
- NEW CONFIG: `max_connections_per_ip` (defaults to 5)
- NEW CONFIG: `moderator_hex_keys` (see next bullet point)
- A rudimentary Management API is now available using https://github.com/nostr-protocol/nips/pull/1325
  To use management front ends against chorus, you must add hex pubkeys to `moderator_hex_keys`.
- Errors about DM kinds are much less common now, as we don't explicitly error unless they specify some set
  of kinds (we implicitly filter out the DMs still)
- Accurate count of bytes sent/received (SSL header data is now counted)
- Kind 10050 dm relay list events are now treated the same way as kind 10002 relay list events.
- Error message detail (e.g. source code line numbers) is now no longer sent to clients.
- Some mild errors are now swallowed.
- Updates of many dependencies, some updates were very large jumps and may change network/http behavior.
- Receipt of a deleted EVENT now returns OK false (was OK true)

# v1.4.0 (2024-05-07, 25058ef4)

- Origin header logged
- CLOSED: auth-required sent if DMs requested and not authenticated yet
- config: minimum_ban_seconds, timeout_seconds, enable_ip_blocking
- default for allow_scrape_if_max_seconds raised from 3600 to 7200
- default for max_subscriptions raised from 32 to 128
- timeouts no longer affect ban seconds
- Internal: switched to pocket for the backend storage
- creates lmdb subdirectory if missing
- several bugfixes: filter parsing, empty tags, event ordering, chorus_compress was fixed

# v1.3.0 (2024-04-10, 7883d754)

- `chorus_moderate` tool to approve events or pubkeys
- `chorus_dump_approvals` tool to view prior approvals
- `chorus_compress` to backup and rebuild your data, compressing as it goes.
- NIP-40 event expiration support
- extended NIP-11 document, with updates to supported NIPs
- Traffic bytes counted and logged on exit (or HUP)
- Logging improvements
- Data migrations. Less space consumed.
- performance improved via 8-byte alignment of events

# v1.2.0 (2024-03-22, a701e148)

- Config option to run as an open relay
- Config variable for scraper behavior
- More efficient scraping
- Reload config on HUP without dropping connections
- IP addresses no longer logged. They are hashed and tracked by their hash.

# v1.1.1 (2024-03-02, fda607a6)

- FIX some LIMIT related bugs
- FIX some indexing bugs that cuased some events that should have been returned to not be.
- Allow scrape up to limit=100, or time range of 1 hour

# v1.1.0 (2024-02-21, ec315d98)

- Update docs

# v1.0.7 (2024-02-21, 4ebe2545)

- BREAKING: Config changed to TOML format
- Logging improvements
- Error handling improvements
- Ban time improvements
- Documentation updates
- Restructured into a `bin` and a `lib`
- Tool to dump all events as JSON to stdout

# v1.0.6 (2024-02-21, 50995a75)

- FIX: REQ errors now properly CLOSED
- Logging improvements
- Persistent IP reputation data for adjusting ban times
- Config settings for `serve_ephmeral` and `serve_relay_lists`
- Uses 'X-Real-Ip' header if behind a proxy

# v1.0.5 (2024-02-20, 5e7c1c38)

- FIX: significant performance problem addressed
- Logging improvements
- EVENTs submitted that are known to be deleted are now responded to with 'blocked:'
- Longer filters now accepted

# v1.0.4 (2024-02-20, 9c7aa299)


src/config.rs
use crate::error::Error;
use hyper::http::uri::{Authority, Scheme, Uri};
use pocket_types::Pubkey;
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use url::Host;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct FriendlyConfig {
    pub data_directory: String,
    pub ip_address: String,
    pub port: u16,
    pub hostname: String,
    pub chorus_is_behind_a_proxy: bool,
    pub base_url: Option<String>,
    pub use_tls: bool,
    pub certchain_pem_path: String,
    pub key_pem_path: String,
    pub name: Option<String>,
    pub description: Option<String>,
    pub banner_url: Option<String>,
    pub icon_url: Option<String>,
    pub privacy_policy: Option<String>,
    pub terms_of_service: Option<String>,
    pub contact: Option<String>,
    #[serde(alias = "public_key_hex")]
    pub contact_public_key_hex: Option<String>,
    pub open_relay: bool,
    pub admin_hex_keys: Vec<String>,
    pub verify_events: bool,
    pub allow_scraping: bool,
    pub allow_scrape_if_limited_to: u32,
    pub allow_scrape_if_max_seconds: u64,
    pub allow_scrape_if_negentropy: bool,
    pub max_subscriptions: usize,
    pub serve_ephemeral: bool,
    pub serve_relay_lists: bool,
    pub server_log_level: String,
    pub library_log_level: String,
    pub client_log_level: String,
    pub enable_ip_blocking: bool,
    pub minimum_ban_seconds: u64,
    pub timeout_seconds: u64,
    pub max_connections_per_ip: usize,
    pub throttling_bytes_per_second: usize,
    pub throttling_burst: usize,
    pub blossom_directory: Option<String>,
    pub enable_negentropy: bool,
}

impl Default for FriendlyConfig {
    fn default() -> FriendlyConfig {
        FriendlyConfig {
            data_directory: "/opt/chorus/var/chorus".to_string(),
            ip_address: "127.0.0.1".to_string(),
            port: 443,
            hostname: "localhost".to_string(),
            chorus_is_behind_a_proxy: false,
            base_url: None,
            use_tls: true,
            certchain_pem_path: "/opt/chorus/etc/tls/fullchain.pem".to_string(),
            key_pem_path: "/opt/chorus/etc/tls/privkey.pem".to_string(),
            name: Some("Chorus Default".to_string()),
            description: Some("A default config of the Chorus relay".to_string()),
            banner_url: None,
            icon_url: None,
            privacy_policy: None,
            terms_of_service: None,
            contact: None,
            contact_public_key_hex: None,
            open_relay: false,
            admin_hex_keys: vec![],
            verify_events: true,
            allow_scraping: false,
            allow_scrape_if_limited_to: 100,
            allow_scrape_if_max_seconds: 7200,
            allow_scrape_if_negentropy: true,
            max_subscriptions: 128,
            serve_ephemeral: true,
            serve_relay_lists: true,
            server_log_level: "Info".to_string(),
            library_log_level: "Info".to_string(),
            client_log_level: "Info".to_string(),
            enable_ip_blocking: true,
            minimum_ban_seconds: 1,
            timeout_seconds: 60,
            max_connections_per_ip: 5,
            throttling_bytes_per_second: 1024 * 1024,
            throttling_burst: 1024 * 1024 * 16,
            blossom_directory: None,
            enable_negentropy: false,
        }
    }
}

impl FriendlyConfig {
    pub fn into_config(self) -> Result<Config, Error> {
        let FriendlyConfig {
            data_directory,
            ip_address,
            port,
            hostname,
            chorus_is_behind_a_proxy,
            base_url,
            use_tls,
            certchain_pem_path,
            key_pem_path,
            name,
            description,
            banner_url,
            icon_url,
            privacy_policy,
            terms_of_service,
            contact,
            contact_public_key_hex,
            open_relay,
            admin_hex_keys,
            verify_events,
            allow_scraping,
            allow_scrape_if_limited_to,
            allow_scrape_if_max_seconds,
            allow_scrape_if_negentropy,
            max_subscriptions,
            serve_ephemeral,
            serve_relay_lists,
            server_log_level,
            library_log_level,
            client_log_level,
            enable_ip_blocking,
            minimum_ban_seconds,
            timeout_seconds,
            max_connections_per_ip,
            throttling_bytes_per_second,
            throttling_burst,
            blossom_directory,
            enable_negentropy,
        } = self;

        let mut contact_public_key: Option<Pubkey> = None;
        if let Some(pkh) = contact_public_key_hex {
            contact_public_key = Some(Pubkey::read_hex(pkh.as_bytes())?);
        };

        let mut admin_keys: Vec<Pubkey> = Vec::with_capacity(admin_hex_keys.len());
        for pkh in admin_hex_keys.iter() {
            admin_keys.push(Pubkey::read_hex(pkh.as_bytes())?);
        }

        let hostname = Host::parse(&hostname)?;

        let server_log_level =
            log::LevelFilter::from_str(&server_log_level).unwrap_or(log::LevelFilter::Info);
        let library_log_level =
            log::LevelFilter::from_str(&library_log_level).unwrap_or(log::LevelFilter::Info);
        let client_log_level =
            log::LevelFilter::from_str(&client_log_level).unwrap_or(log::LevelFilter::Info);

        Ok(Config {
            data_directory,
            ip_address,
            port,
            hostname,
            chorus_is_behind_a_proxy,
            base_url,
            use_tls,
            certchain_pem_path,
            key_pem_path,
            name,
            description,
            banner_url,
            icon_url,
            privacy_policy,
            terms_of_service,
            contact,
            contact_public_key,
            open_relay,
            admin_keys,
            admin_hex_keys,
            verify_events,
            allow_scraping,
            allow_scrape_if_limited_to,
            allow_scrape_if_max_seconds,
            allow_scrape_if_negentropy,
            max_subscriptions,
            serve_ephemeral,
            serve_relay_lists,
            server_log_level,
            library_log_level,
            client_log_level,
            enable_ip_blocking,
            minimum_ban_seconds,
            timeout_seconds,
            max_connections_per_ip,
            throttling_bytes_per_second,
            throttling_burst,
            blossom_directory,
            enable_negentropy,
        })
    }
}

#[derive(Debug, Clone)]
pub struct Config {
    pub data_directory: String,
    pub ip_address: String,
    pub port: u16,
    pub hostname: Host,
    pub chorus_is_behind_a_proxy: bool,
    pub base_url: Option<String>,
    pub use_tls: bool,
    pub certchain_pem_path: String,
    pub key_pem_path: String,
    pub name: Option<String>,
    pub description: Option<String>,
    pub banner_url: Option<String>,
    pub icon_url: Option<String>,
    pub privacy_policy: Option<String>,
    pub terms_of_service: Option<String>,
    pub contact: Option<String>,
    pub contact_public_key: Option<Pubkey>,
    pub open_relay: bool,
    pub admin_keys: Vec<Pubkey>,
    pub admin_hex_keys: Vec<String>,
    pub verify_events: bool,
    pub allow_scraping: bool,
    pub allow_scrape_if_limited_to: u32,
    pub allow_scrape_if_max_seconds: u64,
    pub allow_scrape_if_negentropy: bool,
    pub max_subscriptions: usize,
    pub serve_ephemeral: bool,
    pub serve_relay_lists: bool,
    pub server_log_level: log::LevelFilter,
    pub library_log_level: log::LevelFilter,
    pub client_log_level: log::LevelFilter,
    pub enable_ip_blocking: bool,
    pub minimum_ban_seconds: u64,
    pub timeout_seconds: u64,
    pub max_connections_per_ip: usize,
    pub throttling_bytes_per_second: usize,
    pub throttling_burst: usize,
    pub blossom_directory: Option<String>,
    pub enable_negentropy: bool,
}

impl Default for Config {
    fn default() -> Config {
        let friendly = FriendlyConfig::default();

        // We know the default config passes into_config without error:
        friendly.into_config().unwrap()
    }
}

impl Config {
    /// Get the URI for our server matching the inner Uri, overridden with either
    /// our base_url parts or our hostname/port.
    pub fn uri_parts(&self, inner: Uri, http: bool) -> Result<http::uri::Parts, Error> {
        let mut uri_parts = inner.into_parts();

        if let Some(s) = &self.base_url {
            let base_uri = s.parse::<Uri>()?;
            let base_uri_parts = base_uri.into_parts();
            uri_parts.scheme = base_uri_parts.scheme;
            uri_parts.authority = base_uri_parts.authority;
        } else {
            let scheme = match (self.use_tls, http) {
                (false, false) => Scheme::from_str("ws").unwrap(),
                (true, false) => Scheme::from_str("wss").unwrap(),
                (false, true) => Scheme::HTTP,
                (true, true) => Scheme::HTTPS,
            };
            uri_parts.scheme = Some(scheme);

            let authority = Authority::from_str(&format!("{}:{}", self.hostname, self.port))?;
            uri_parts.authority = Some(authority);
        }

        Ok(uri_parts)
    }
}
